### PR TITLE
Add upload to registry workflow for foundry_std

### LINKED ISSUE #2463: Add publish workflow for snforge_scarb_plugin
### Which components does the task require to be changed? (think hard pls)

other (describe below)

### Description

Related to https://github.com/foundry-rs/starknet-foundry/issues/2344

Under said issue automatic uploads for snforge_std and sncast_std was added. At the time of creating relevant PRs there was still no support for scarb package for procmacros. The aim of this issue is to update the publish workflow, to also automatically upload new snforge_scarb_plugin whenever its version changes. This could probably be a separate publish flow for each push to master, where `crates/snforge_scarb_plugin/Scarb.toml` is changed and its version is not already deployed to the registry.

Blocked by https://github.com/software-mansion/scarb/issues/1527

Up until now, whenever a new version of snforge_scarb_plugin is needed please construct the package manually. If needed, seek help from scarb team.

### PR DESCRIPTION
**Pull request by THenry14**
Closes https://github.com/foundry-rs/starknet-foundry/issues/2344
Closes #2463 

This PR is a part of the stack:
-- Add missing info recommended for package uploads (https://github.com/foundry-rs/starknet-foundry/pull/2460)
--> Add github action workflow for automatic registry uploads during release (This PR)

## Introduced changes

- Updates github workflow to also upload std libraries to scarbs registry during release
- Sets snforge_scarb_plugin to specific version downloadable from the registry, as required by scarb

## Checklist

- [X] Linked relevant issue
- [X] Updated relevant documentation
- [X] Added relevant tests
- [X] Performed self-review of the code
- [X] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by franciszekjob** [comment]
Blocked by https://github.com/software-mansion/scarb/pull/1605

**Comment by kkawula** [comment]
It needs clarification, but plugins weren't correctly handled by `scarb` until https://github.com/software-mansion/scarb/pull/1712 was merged, I suppose that it will be needed to update `scarb` version in the entire project. 
In the earlier versions `scarb` had problem with access to the cache in database which caused dead lock. 
CI to lookup: https://github.com/foundry-rs/starknet-foundry/actions/runs/11577774705

**Comment by maciektr** [comment]
> It needs clarification, but plugins weren't correctly handled by scarb until https://github.com/software-mansion/scarb/pull/1712 was merged, I suppose that it will be needed to update scarb version in the entire project.

It shouldn't have anything to do with uploading packages though

File: .github/workflows/publish_packages.yml
Comment (kkawula) [inline .github/workflows/publish_packages.yml:35]: It gives cool info on what was published, in the sense that we do not want update the plugin every release.
image

File: .github/workflows/publish_packages.yml
Comment (kkawula) [inline .github/workflows/publish_packages.yml:51]: `--allow-dirty` because we modify `snforge_std` by executing the script without committing changes

File: .github/workflows/publish_packages.yml
Comment (kkawula) [inline .github/workflows/publish_packages.yml:24]: created issue https://github.com/foundry-rs/starknet-foundry/issues/2669

**Comment by kkawula** [comment]
it works, tested on dev registry.
- [workflow](https://github.com/kkawula/starknet-foundry/actions/runs/11815728427/job/32917498713)
- [dev_registry](https://scarbs.dev/packages/snforge_std/0.33.0)

File: .github/workflows/publish_packages.yml
Comment (DelevoXDG) [inline .github/workflows/publish_packages.yml:26]: Is this step necessary?

We implemented this for publishing Scarb plugins because their version matches the Cairo version. Since multiple Scarb versions can correspond to a single Cairo version, running the workflow with every Scarb release could cause errors by publishing plugins with the same version. 

Afaik, this doesn’t apply to snfoundry plugins because their version isn’t tied to any Cairo version. Normally, the workflow should only be triggered once per tag.

File: .github/workflows/publish_packages.yml
Comment (DelevoXDG) [inline .github/workflows/publish_packages.yml:6]: I think we only want to publish on tag release

**Review by DelevoXDG — COMMENTED** [review]
Partial review, please re-request when ready 👍🏻

File: .github/workflows/publish_packages.yml
Comment (cptartur) [inline .github/workflows/publish_packages.yml:5]: I think `published` will also trigger on creating a draft release which our workflow does. I think it should be `released` so it only uploads to registry if we publish a new release.

Comment (cptartur) [reply .github/workflows/publish_packages.yml:5]: Also a concern, in case we ever create a pre-release, do we want to upload to registry as well? Does it support pre-release versions?

cc @maciektr

File: .github/workflows/publish_packages.yml
Comment (cptartur) [reply .github/workflows/publish_packages.yml:6]: I think there might be cases where we want to trigger upload manually. I think we can keep it

File: .github/workflows/publish_packages.yml
Comment (cptartur) [reply .github/workflows/publish_packages.yml:26]: We might decouple foundry plugin version from foundry itself so it might be the case that for consecutive foundry releases, plugin version will stay the same.

So we have to make sure we don't reupload the plugin to the registry in these cases.

Comment (kkawula) [reply .github/workflows/publish_packages.yml:26]: Sometimes we do patch plugin, without release then we would like to publish it and sometimes we will not do release plugin together with snforge_std and sncast_std, we want to set it on 1.0.0 and keep it until there are changes, and then I would fail the workflow and CI will be red at the release

File: .github/workflows/publish_packages.yml
Comment (cptartur) [inline .github/workflows/publish_packages.yml:47]: Nit, but maybe this is more readable?
```suggestion
        if: steps.check-plugins.outputs.snforge_scarb_plugin == 'false'
```

File: scripts/set_plugin_version.sh
Comment (cptartur) [inline scripts/set_plugin_version.sh:5]: Assuming this is meant to extract the version, why not use this code again?

```shell
snforge_std_version=$(cat snforge_std/Scarb.toml | grep version | cut -d '"' -f 2
```

File: scripts/set_plugin_version.sh
Comment (cptartur) [inline scripts/set_plugin_version.sh:11]: If we are always removing the `.bak` file anyway, why not just omit creating it?

**Review by cptartur — CHANGES_REQUESTED** [review]

File: .github/workflows/publish_packages.yml
Comment (kkawula) [reply .github/workflows/publish_packages.yml:5]: You are right, it should be changed to `released`
https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=released#release

File: .github/workflows/publish_packages.yml
Comment (DelevoXDG) [reply .github/workflows/publish_packages.yml:26]: As I pointed out offline, if you're going in the direction of decoupling plugin from foundry itself, it seems to me it a separate workflow should be considered

File: .github/workflows/publish_packages.yml
Comment (kkawula) [reply .github/workflows/publish_packages.yml]: done

File: .github/workflows/publish_packages.yml
Comment (DelevoXDG) [reply .github/workflows/publish_packages.yml:57]: Why not require from tag that triggers the workflow to already include changes introduced by `set_plugin_version.sh`? 🤔

Comment (kkawula) [reply .github/workflows/publish_packages.yml:57]: For development reason, we always would like to have a relative path specified in the dependency

File: .github/workflows/publish_packages.yml
Comment (kkawula) [reply .github/workflows/publish_packages.yml:26]: It is one of the opportunities, but isn't it good enough? @cptartur

Comment (cptartur) [reply .github/workflows/publish_packages.yml:26]: As on how the decoupling will go exactly is yet to be determined. I think the workflow in the current form is okay, and we can introduce changes in the future.

File: .github/workflows/publish_packages.yml
Comment (cptartur) [reply .github/workflows/publish_packages.yml]: Okay, if we want to upload pre-releases as well, let's keep it at published.

File: .github/workflows/publish_packages.yml
Comment (cptartur) [inline .github/workflows/publish_packages.yml:5]: Following the discussion from different comment 

```suggestion
      - published
```

File: .github/workflows/publish_packages.yml
Comment (cptartur) [inline .github/workflows/publish_packages.yml:48]: Actually, shouldn't it be without quotes in case of booleans? So just `false`? 

I've checked in the GH docs and it's not 100% clear to me.

Comment (cptartur) [reply .github/workflows/publish_packages.yml:48]: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions

https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions

**Review by cptartur — COMMENTED** [review]
@THenry14 Can you please review as well? I know it's technically your PR but it's changed a bit 😅

File: .github/workflows/publish_packages.yml
Comment (THenry14) [reply .github/workflows/publish_packages.yml:26]: Similarly to @DelevoXDG , I would probably be in favour of splitting release flows for macros and standard libraries as well. 

My understanding is that the plugin won't necessarily follow the *_std version, so it may make sense to release it independently. For this manual workflow dispatch only is good enough. Most of logic from this file is fine, though you wouldn't need anything from this step (I mean the check-versions, which I find very hacky) so I'd actually argue it could become far simpler.

Having this, we could probably incorporate *_std upload to the `release.yml` workflow, as it seems to me this is more natural place for this. We already verify the version there, so chances are we do not need any further modifications - we could probably just add a new job (standard-libraries-release, or scarbs-release or sth) that will set the plugin version (this is unnecessary if you decide to set the version in Scarb.toml in repo) and use scarb to publish and call it a day. No need to check versions against the registry, verify what should/should not be published etc. 

Thoughts?

File: scripts/set_plugin_version.sh
Comment (THenry14) [inline scripts/set_plugin_version.sh:5]: useless use of cat

https://www.shellcheck.net/wiki/SC2002

File: .github/workflows/publish_packages.yml
Comment (THenry14) [reply .github/workflows/publish_packages.yml:57]: Is it a hard requirement to set the version in snforge only for uploads? What are the downsides of having this set in repo?

File: scripts/set_plugin_version.sh
Comment (THenry14) [reply scripts/set_plugin_version.sh:5]: This probably explains it better
https://en.wikipedia.org/wiki/Cat_(Unix)#Useless_use_of_cat

File: .github/workflows/publish_packages.yml
Comment (kkawula) [reply .github/workflows/publish_packages.yml:26]: I generally agree with you, but I can still see a few problems we could encounter. In scenario where this workflow is splited and moved to `release` we can get error while releasing new versions of _std and plugins simultaneously. Then plugin should be then uploaded before _std, to make `snforge_std` dependency working. As other job in can be made sequential, but if we like to provide workflow_dispatcher then its generates a lot of unnecessary boilerplate code (`ifs`) in entire workflow, because we can’t actually specify exact job to run on dispatch. In scenario of two different workflows it actually can’t be made sequential and then we can only hope that plugin will be uploaded first.

File: .github/workflows/publish_packages.yml
Comment (kkawula) [reply .github/workflows/publish_packages.yml:57]: Yes, we would like always to test the latest plugin version while developing, so relative path has to be provided. e.g. e2e test
There is always a case where it should be changed this way or another, definitely here will be the easiest.

File: scripts/set_plugin_version.sh
Comment (kkawula) [reply scripts/set_plugin_version.sh:5]: Cool, removed

File: .github/workflows/publish_packages.yml
Comment (THenry14) [reply .github/workflows/publish_packages.yml:26]: > In scenario where this workflow is splited and moved to release we can get error while releasing new versions of _std and plugins simultaneously

Well I do not think we necessarily need to release macro in the same time we release package, and because of that workflow_dispatch would be ideal. I can easily imagine macro not being updated for months, or be updated multiple times during one week when we do not want to release new snfoundry version.

If you wanted to the release workflow could call the plugin publish workflow, to have the functionality we're aiming for and be sure that the release operations are indeed sequential.

Comment (DelevoXDG) [reply .github/workflows/publish_packages.yml:26]: > If you wanted to the release workflow could call the plugin publish workflow, to have the functionality we're aiming for and be sure that the release operations are indeed sequential.

That's how I see it as well, we should have:
1. Plugin publish workflow with `workflow_dispatch` and `workflow_call`
2. Foundry publish workflow on `release` and `workflow_dispatch`. It should first should check if plugin is up to date in first job and re-use the first workflow if necessary to upload the plugin, and release the foundry libs afterwards

> As other job in can be made sequential, but if we like to provide workflow_dispatcher then its generates a lot of unnecessary boilerplate code (`ifs`) in entire workflow, because we can’t actually specify exact job to run on dispatch.

Do you have any specific examples of why that might be the case? 🤔 I don't think this should really cause any boilerplate or significantly more ifs than we have now

Comment (kkawula) [reply .github/workflows/publish_packages.yml]: Seems convincing. Updated.

File: .github/workflows/release.yml
Comment (THenry14) [inline .github/workflows/release.yml:265]: Are you sure this curl works as intended? I'm getting:

```
$ curl -s https://scarbs.xyz/api/v1/index/sn/fo/snforge_scarb_plugin.json | jq --arg version 0.31.1 '[.[] | select(.v == $version)] | length > 0'
false

$ curl -s https://scarbs.xyz/api/v1/index/sn/fo/snforge_scarb_plugin.json | jq --arg version 31.31.1 '[.[] | select(.v == $version)] | length > 0'
false
```

So this returns false not matter if the version exists or not. Maybe I messed something though?

File: .github/workflows/release.yml
Comment (THenry14) [inline .github/workflows/release.yml:266]: nit: you can add -x to set flags for this script, and get rid of this echo (as you'll be able to see the checked version during the jq check)

File: .github/workflows/release.yml
Comment (THenry14) [inline .github/workflows/release.yml:291]: the same as in plugin workflow

File: scripts/set_plugin_version.sh
Comment (THenry14) [inline scripts/set_plugin_version.sh:7]:
```suggestion
PLUGIN_FILE_PATH="../crates/snforge-scarb-plugin/Scarb.toml"
STD_FILE_PATH="../snforge_std/Scarb.toml"

VERSION=$(grep version "$PLUGIN_FILE_PATH" | cut -d '"' -f 2)
```
nitpick

btw https://en.wikipedia.org/wiki/STD_(disambiguation) ( ͡° ͜ʖ ͡°)

Let's maybe use full `SNFORGE_STD_PATH` name

**Review by THenry14 — COMMENTED** [review]
great job mate

File: .github/workflows/release.yml
Comment (cptartur) [inline .github/workflows/release.yml:254]: Should probably add `needs: [ create-release ]` so it only triggers on successful release.

File: .github/workflows/release.yml
Comment (cptartur) [inline .github/workflows/release.yml:277]:
```suggestion
    needs: [ create-release ]
```

How about this, I think then it will only run on successfully created release.

File: .github/workflows/release.yml
Comment (DelevoXDG) [inline .github/workflows/release.yml:253]: Nit: 
```suggestion
  upload-plugin-version:
```
would be more accurate

File: .github/workflows/release.yml
Comment (DelevoXDG) [reply .github/workflows/release.yml:265]: It works, there's just no version `0.31.1` in the registry 😄

Comment (DelevoXDG) [reply .github/workflows/release.yml:265]: Nit: 
`snforge_scarb_plugin_exists`, `snforge_scarb_plugin_uploaded` or `snforge_scarb_plugin_published` would be most accurate since we're dealing with true or false statement

File: .github/workflows/publish_plugin.yml
Comment (DelevoXDG) [inline .github/workflows/publish_plugin.yml:1]:
```suggestion
name: Publish snforge_scarb_plugin
```

File: .github/workflows/release.yml
Comment (DelevoXDG) [inline .github/workflows/release.yml:274]: Nit:
```suggestion
  publish-to-registry:
```
(for consistency)

File: .github/workflows/release.yml
Comment (DelevoXDG) [reply .github/workflows/release.yml:277]: > How about this, I think then it will only run on successfully created release.

Yes, but we still need to ensure that latest plugin version is released before publishing the stds.  

https://github.com/foundry-rs/starknet-foundry/pull/2461#discussion_r1847852740

File: .github/workflows/release.yml
Comment (THenry14) [reply .github/workflows/release.yml:265]: This is what I am

obraz

File: .github/workflows/publish_plugin.yml
Comment (DelevoXDG) [inline .github/workflows/publish_plugin.yml:34]: https://github.com/foundry-rs/starknet-foundry/pull/2461/files#r1853809134

This should also always run on `workflow_dispatch`

**Review by DelevoXDG — APPROVED** [review]
Please address the comments before merging

File: .github/workflows/release.yml
Comment (cptartur) [inline .github/workflows/release.yml:277]: I don't think we need that here
```suggestion
        run: scarb publish
```

**Review by cptartur — APPROVED** [review]

File: .github/workflows/release.yml
Comment (kkawula) [reply .github/workflows/release.yml:277]: For unknown reasons, I got a warning about uncommitted changes if `workflow_call` was triggered. Leaving it as it is

**Pull request merged**
